pkg/ddl/backfilling.go

```diff
@@ -691,53 +597,120 @@ func (dc *ddlCtx) writePhysicalTableRecord(sessPool *sess.Pool, t table.Physical
 	failpoint.Inject("MockCaseWhenParseFailure", func(val failpoint.Value) {
 		//nolint:forcetypeassert
 		if val.(bool) {
-			failpoint.Return(errors.New("job.ErrCount:" + strconv.Itoa(int(job.ErrorCount)) + ", mock unknown type: ast.whenClause."))
+			failpoint.Return(errors.New("job.ErrCount:" + strconv.Itoa(int(reorgInfo.Job.ErrorCount)) + ", mock unknown type: ast.whenClause."))
 		}
 	})
 
 	jc := reorgInfo.NewJobContext()
-	sessCtx := newContext(reorgInfo.d.store)
-	scheduler, err := newBackfillScheduler(dc.ctx, reorgInfo, sessPool, bfWorkerType, t, sessCtx, jc)
+	sessCtx := newReorgSessCtx(reorgInfo.d.store)
+
+	eg, egCtx := util.NewErrorGroupWithRecoverWithCtx(dc.ctx)
+
+	scheduler, err := newBackfillScheduler(egCtx, reorgInfo, sessPool, bfWorkerType, t, sessCtx, jc)
 	if err != nil {
 		return errors.Trace(err)
 	}
 	defer scheduler.close(true)
-
-	consumer := newResultConsumer(dc, reorgInfo, sessPool, false)
-	consumer.run(scheduler, startKey, &totalAddedCount)
-
 	err = scheduler.setupWorkers()
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	taskIDAlloc := newTaskIDAllocator()
-	for {
-		kvRanges, err := splitTableRanges(t, reorgInfo.d.store, startKey, endKey, backfillTaskChanSize)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		if len(kvRanges) == 0 {
-			break
-		}
-		logutil.BgLogger().Info("start backfill workers to reorg record", zap.String("category", "ddl"),
-			zap.Stringer("type", bfWorkerType),
-			zap.Int("workerCnt", scheduler.currentWorkerSize()),
-			zap.Int("regionCnt", len(kvRanges)),
-			zap.String("startKey", hex.EncodeToString(startKey)),
-			zap.String("endKey", hex.EncodeToString(endKey)))
-
-		sendTasks(scheduler, consumer, t, kvRanges, reorgInfo, taskIDAlloc)
-		if consumer.shouldAbort() {
-			break
+	// process result goroutine
+	eg.Go(func() error {
+		totalAddedCount := reorgInfo.Job.GetRowCount()
+		keeper := newDoneTaskKeeper(startKey)
+		cnt := 0
+
+		for {
+			select {
+			case <-egCtx.Done():
+				return egCtx.Err()
+			case result, ok := <-scheduler.resultChan():
+				if !ok {
+					ddlLogger.Info("backfill workers successfully processed",
+						zap.Stringer("element", reorgInfo.currElement),
+						zap.Int64("total added count", totalAddedCount),
+						zap.String("start key", hex.EncodeToString(startKey)))
+					return nil
+				}
+				cnt++
+
+				if result.err != nil {
+					ddlLogger.Warn("backfill worker failed",
+						zap.Int64("job ID", reorgInfo.ID),
+						zap.Int64("total added count", totalAddedCount),
+						zap.String("start key", hex.EncodeToString(startKey)),
+						zap.String("result next key", hex.EncodeToString(result.nextKey)),
+						zap.Error(result.err))
```

Comment:
Do we need to cancel the ctx to notify we need stop `sendTasks` ?

Reply:
`eg` is a wrapper of `errgroup` package, it can automatically cancel the context after `Go` returns an error

> // The derived Context is canceled the first time a function passed to Go
// returns a non-nil error or the first time Wait returns, whichever occurs
// first.

And `eg.Wait` returns the first error of `eg.Go`, it simplifies our logic a lot.